File: Data Structures and Algorithms/4. Recursion/Kunal/Arrays/linearSearch.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> l;

bool linear_search(vector<int>& a, int target, int index) {
    if(index == a.size())
        return false;

    return (a[index] == target) or linear_search(a, target, index + 1);
}

void find_all_index(vector<int>& a, int target, int index) {
    if(index == a.size())
        return;
    if(a[index] == target)
        l.push_back(index);
    find_all_index(a, target, index + 1);
}

int findIndex(vector<int>& a, int target, int index) {
    if(index == a.size())
        return -1;
    if(a[index] == target)
        return index;
    else
        return findIndex(a, target, index + 1);
}

int main(){
    int n;	cin >> n;
    vector<int> a(n);
    for(int i = 0; i < n; i++){
        cin >> a[i];
    }

    int target;
    cin >> target;
    int index = findIndex(a, target, 0);
    cout << target << " found at index: " << index << '\n';

    find_all_index(a, target, 0);
    for(int i = 0; i < l.size(); i++){
        cout << l[i] << " ";
    }

    return 0;
}

File: Data Structures and Algorithms/4. Recursion/Kunal/Arrays/LinearSearch.java
import java.util.*;
public class LinearSearch {
    public static boolean present(int[] arr, int target, int index) {
        if(index == arr.length)
            return false;
        return arr[index] == target || present(arr, target, index + 1);
    }

    public static int findIndex(int[] arr, int target, int index) {
        // base condition
        if(index == arr.length)
            return -1;
        // body
        if(arr[index] == target)
            return index;
        else
        // recurrence relation
            return findIndex(arr, target, index + 1);
    }

    public static int findIndexLast(int[] arr, int target, int index) {
        // base condition
        if(index == -1)
            return -1;
        // body
        if(arr[index] == target)
            return index;
        else
        // recurrence relation
            return findIndex(arr, target, index - 1);
    }
    
    public static ArrayList findAllIndex(int[] arr, int target, int index, ArrayList<Integer> list) {
        if(index == arr.length)
            return list;
        if(arr[index] == target)
            list.add(index);
        return findAllIndex(arr, target, index + 1, list);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int []arr = new int[n];
        for(int i = 0; i < n; i++){
            arr[i] = sc.nextInt();
        }
        int target = sc.nextInt();
        boolean test = present(arr, target, 0);
        if(test) {
            System.out.print("present.\n");
        } else {
            System.out.print("not present.\n");
        }

        int first = findIndex(arr, target, 0);
        int last = findIndexLast(arr, target, arr.length - 1);
        System.out.print(target + " Found at index: " + first + "\n");
        System.out.print(target + " Found at index: " + last + "\n");

        System.out.print(findAllIndex(arr, target, 0, new ArrayList<>()));
    }
}
